File: backend/src/audit.js
import { prisma } from "./prisma.js";

const HIDDEN_KEYS = new Set(["password", "passwordHash", "authToken", "keyPem", "certPem", "token"]);

function sanitizeValue(value) {
  if (value == null) return value;
  if (Array.isArray(value)) return value.slice(0, 20).map((x) => sanitizeValue(x));
  if (typeof value === "object") {
    const out = {};
    for (const [k, v] of Object.entries(value)) {
      out[k] = HIDDEN_KEYS.has(k) ? "[redacted]" : sanitizeValue(v);
    }
    return out;
  }
  if (typeof value === "string" && value.length > 400) return `${value.slice(0, 400)}…`;
  return value;
}

function deriveTarget(path, params = {}) {
  if (path.includes("/agent/nodes/")) {
    return { targetType: "agent-node", targetId: params.id || params.nodeId || null };
  }
  if (path.includes("/vpn-users/")) {
    return { targetType: "vpn-user", targetId: params.id || null };
  }
  if (path.includes("/organizations/")) {
    return { targetType: "organization", targetId: params.id || null };
  }
  if (path.includes("/certificates/")) {
    return { targetType: "certificate", targetId: params.id || null };
  }
  if (path.includes("/admins/")) {
    return { targetType: "admin", targetId: params.id || null };
  }
  return { targetType: null, targetId: null };
}

function pickAction(req) {
  const p = req.path || "";
  if (p.includes("/openvpn/service")) return "openvpn-service-action";
  if (p.includes("/openvpn-settings")) return "openvpn-settings-update";
  if (p.includes("/disconnect")) return "vpn-client-disconnect";
  return `${req.method.toLowerCase()} ${p}`;
}

function clientIp(req) {
  const xff = req.headers["x-forwarded-for"];
  if (typeof xff === "string" && xff) return xff.split(",")[0].trim();
  if (Array.isArray(xff) && xff.length > 0) return String(xff[0]).trim();
  return req.ip || req.socket?.remoteAddress || "";
}

export function auditAdminActions(req, res, next) {
  if (!["POST", "PUT", "PATCH", "DELETE"].includes(req.method)) return next();
  if (!req.user?.sub) return next();

  const startedAt = new Date();
  const action = pickAction(req);
  const body = sanitizeValue(req.body);

  res.on("finish", async () => {
    try {
      const { targetType, targetId } = deriveTarget(req.path, req.params);
      await prisma.adminActionLog.create({
        data: {
          adminId: String(req.user.sub),
          adminUsername: String(req.user.username || ""),
          method: req.method,
          path: req.originalUrl || req.path,
          action,
          targetType,
          targetId,
          ipAddress: clientIp(req),
          userAgent: req.headers["user-agent"] ? String(req.headers["user-agent"]) : null,
          statusCode: Number(res.statusCode || 0),
          details: body || null,
          createdAt: startedAt,
        },
      });
    } catch (error) {
      console.error("Failed to write admin action log:", error.message);
    }
  });

  next();
}

File: backend/internal/audit/audit.go
package audit

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/segmentio/ksuid"

	"github.com/openvpn-control/openvpn-control-server/backend/internal/auth"
)

var hiddenKeys = map[string]struct{}{
	"password": {}, "passwordHash": {}, "authToken": {}, "keyPem": {}, "certPem": {}, "token": {},
}

type responseRecorder struct {
	http.ResponseWriter
	status int
}

func (r *responseRecorder) WriteHeader(code int) {
	r.status = code
	r.ResponseWriter.WriteHeader(code)
}

// Middleware logs mutating admin API calls to AdminActionLog (Node auditAdminActions).
func Middleware(pool *pgxpool.Pool) func(http.Handler) http.Handler {
	disabled := os.Getenv("AUDIT_ADMIN_ACTIONS_DISABLED") == "1"
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if disabled {
				next.ServeHTTP(w, r)
				return
			}
			switch r.Method {
			case http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete:
			default:
				next.ServeHTTP(w, r)
				return
			}
			claims := auth.UserFromRequest(r)
			if claims == nil || claims.Sub == "" {
				next.ServeHTTP(w, r)
				return
			}

			startedAt := time.Now()
			var bodyCopy []byte
			if r.Body != nil && r.ContentLength != 0 {
				bodyCopy, _ = io.ReadAll(r.Body)
				r.Body = io.NopCloser(bytes.NewReader(bodyCopy))
			}
			details := sanitizeJSON(bodyCopy)

			rec := &responseRecorder{ResponseWriter: w, status: http.StatusOK}
			next.ServeHTTP(rec, r)

			path := r.URL.Path
			if q := r.URL.RawQuery; q != "" {
				path = path + "?" + q
			}
			targetType, targetID := deriveTarget(r)
			go func() {
				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
				defer cancel()
				detailsJSON, _ := json.Marshal(details)
				var detailsArg any
				if len(detailsJSON) > 2 {
					detailsArg = detailsJSON
				}
				_, _ = pool.Exec(ctx, `
					INSERT INTO "AdminActionLog" (
						id, "adminId", "adminUsername", method, path, action,
						"targetType", "targetId", "ipAddress", "userAgent", "statusCode", details, "createdAt"
					) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)`,
					ksuid.New().String(),
					claims.Sub,
					claims.Username,
					r.Method,
					path,
					pickAction(r),
					targetType,
					targetID,
					clientIP(r),
					strings.TrimSpace(r.UserAgent()),
					rec.status,
					detailsArg,
					startedAt,
				)
			}()
		})
	}
}

func pickAction(r *http.Request) string {
	p := r.URL.Path
	if strings.Contains(p, "/openvpn/service") {
		return "openvpn-service-action"
	}
	if strings.Contains(p, "/openvpn-settings") {
		return "openvpn-settings-update"
	}
	if strings.Contains(p, "/disconnect") {
		return "vpn-client-disconnect"
	}
	return strings.ToLower(r.Method) + " " + p
}

func deriveTarget(r *http.Request) (*string, *string) {
	p := r.URL.Path
	id := chi.URLParam(r, "id")
	if id == "" {
		id = chi.URLParam(r, "nodeId")
	}
	var tt, tid string
	switch {
	case strings.Contains(p, "/agent/nodes/"), strings.Contains(p, "/panel/nodes/"):
		tt, tid = "agent-node", id
	case strings.Contains(p, "/vpn-users/"):
		tt, tid = "vpn-user", id
	case strings.Contains(p, "/organizations/"):
		tt, tid = "organization", id
	case strings.Contains(p, "/certificates/"):
		tt, tid = "certificate", id
	case strings.Contains(p, "/admins/"):
		tt, tid = "admin", id
	default:
		return nil, nil
	}
	if tid == "" {
		return &tt, nil
	}
	return &tt, &tid
}

func clientIP(r *http.Request) string {
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		if i := strings.Index(xff, ","); i >= 0 {
			return strings.TrimSpace(xff[:i])
		}
		return strings.TrimSpace(xff)
	}
	host, _, _ := strings.Cut(r.RemoteAddr, ":")
	return host
}

func sanitizeJSON(raw []byte) any {
	if len(raw) == 0 {
		return nil
	}
	var v any
	if json.Unmarshal(raw, &v) != nil {
		return nil
	}
	return sanitizeValue(v)
}

func sanitizeValue(v any) any {
	switch t := v.(type) {
	case nil:
		return nil
	case []any:
		if len(t) > 20 {
			t = t[:20]
		}
		out := make([]any, len(t))
		for i, x := range t {
			out[i] = sanitizeValue(x)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, val := range t {
			if _, hide := hiddenKeys[k]; hide {
				out[k] = "[redacted]"
			} else {
				out[k] = sanitizeValue(val)
			}
		}
		return out
	case string:
		if len(t) > 400 {
			return t[:400] + "…"
		}
		return t
	default:
		return t
	}
}
